torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -69,11 +69,14 @@ class CpuTimer : public Timer {
 
 C10_REGISTER_TYPED_CLASS(TimerRegistry, c10::kCPU, CpuTimer);
 
-bool is_lazy_or_alias(const at::Tensor& tensor, const at::Tensor& view) {
-  if (tensor.device().type() == at::kLazy) {
-    // TODO(alanwaketan): Currently we can't assert this as bucket_view_out (L1457) is on cuda given
-    // it's set by ProcessGroupNCCL::collective (L1480) which is before the copyLazy hack.
-    // TORCH_INTERNAL_ASSERT(view.device().type() == at::kLazy);
+bool is_alias(const at::Tensor& tensor, const at::Tensor& view) {
+  // Some TensorImpl subclasses might not have storages, for example,
+  // LazyTensor (LTCTensorImpl). Thus, we double check that before
+  // calling into tensor.is_alias_of() which relies on the storages
+  // to perform the actual check.
+  // FIXME(alanwaketan): 1) Investigate why LazyTensor demands this to
+  // return true. 2) Consider moving this check to tensor.is_alias_of().
+  if (!tensor.has_storage() || !view.has_storage()) {
     return true;
```

Comment:
Ideally, we should return false here. However, LazyTensor somehow demands gradient_as_bucket_view to be set, which we then need to return true here to comply with that. Need more time to investigate why.

Reply:
this is the part i understand least, could you explain the reasoning?

what I understand is, lazy tensors don't have storage, so we can't call  lazy_tensor.is_alias_of(view).

what I don't understand is, if we determine tensor or view has no storage, all that implies is that it could be lazy or some other special type.  

it seems like we may return true here when tensor and view are not actually aliases.

what is the importance of returning true or false here- is it 'safe' to return true (it just prevents some bucketing optimization from happening?), while it is 'unsafe' to return false (it could cause an unsafe optimization to happen?)

Reply:
It's unsafe no matter whether this function returns true or false for unrecognized tensor type. I guess a better fix could as I suggest is to modify tensor.is_alias_of() to accommodate extended tensor type. The current implementation is more like a hack.